src/components/contract-boost-calculator.tsx

```diff
@@ -1,484 +1,632 @@
-// this lint rule is completely accurate but i cba to deal with it right now
-/* eslint-disable react/jsx-no-bind */
-
-import * as Accordion from '@radix-ui/react-accordion';
-import { ChevronDownIcon } from '@radix-ui/react-icons';
-import * as Tooltip from '@radix-ui/react-tooltip';
-import { useEffect, useState, useCallback } from 'react';
+import { useCallback, createContext, useContext, useMemo, useEffect, useRef } from 'react';
 import type { ChangeEvent } from 'react';
-import CustomTextInput, { CustomNumberInput, CustomSelectInput } from '/components/CustomInput';
+import type { Artifact, ArtifactIntensity, Stone } from '/lib/ei_data';
 import {
 	artifactRarity,
-	BASE_MAX_HAB_SPACE,
-	boostSetPresets,
-	chaliceOptions,
-	formInitialState,
-	gussetOptions,
-	monocleOptions,
-	stonesCountOptions,
-} from '/lib/data/contract-boost-constants';
-
-type EquippedArtifact = {
-	spec: {
-		// what
-		level: number;
-		name: number;
-		rarity: number;
+	ArtifactSpec,
+	PossibleArtifactIntensity,
+	chaliceMultiplier,
+	monocleMultiplier,
+	gussetMultiplier,
+} from '/lib/ei_data';
+import { Boost } from '/components/Boosts';
+import { useToggleState } from '/lib/utils';
+import { generateCalculator, type WithSetter } from '/components/calculator.tsx';
+
+type SlottedArtifact = {
+	spec: Artifact;
+	stonesList: Stone[];
+};
+type EIBackupResponse = {
+	userName: string;
+	artifactsDb: {
+		savedArtifactSetsList: Array<{
+			slotsList: Array<{
+				occupied: boolean;
+				itemId: number;
+			}>;
+		}>;
+		inventoryItemsList: Array<{
+			itemId: number;
+			artifact: SlottedArtifact;
+		}>;
 	};
 };
-export default function ContractBoostCalculator() {
-	const [formState, setFormState] = useState(() => {
-		const savedFormState = typeof window === 'undefined' ? null : localStorage.getItem('formState');
-		return savedFormState ? JSON.parse(savedFormState) : formInitialState;
-	});
-
-	const [equippedArtifactsByIGN, setEquippedArtifactsByIGN] = useState([]);
-	const [equippedArtifactsListForSelectedIGN, setEquippedArtifactsListForSelectedIGN] = useState(
-		{} as any,
-	);
+const ApiUriContext = createContext<string>('missing api uri');
+
+const enum FetchState {
+	IDLE = 'idle',
+	PENDING = 'pending',
+	RETRY = 'retry',
+	SUCCESS = 'success',
+	FAILURE = 'failure',
+}
 
-	const [selectedBoostPreset, setSelectedBoostPreset] = useState('1');
-	const [availableIGNOptions, setAvailableIGNOptions] = useState([
-		{ text: 'Select an option', value: '' },
-	]);
+type CalcData = {
+	eid: string;
+	gusset?: ArtifactIntensity | undefined;
+	monocle?: ArtifactIntensity | undefined;
+	chalice?: ArtifactIntensity | undefined;
```

Comment:
use `?` or `undefined`

Reply:
our tsconfig sets (via `astro/tsconfigs/strictest`) `exactOptionalPropertyTypes`, which means that without `?` we can't leave these keys out, which I want, and without `| undefined`, we can't pass literal `undefined`, which I need:

https://www.typescriptlang.org/play/?exactOptionalPropertyTypes=true#code/C4TwDgpgBAIhC2B7KBeKBvAUFKAnCAjgK4CW+AJgFxQDOwuJAdgOYDc2UxEdJijA-NToMW7HEUbkIAM2kQqtek2ZQAPlAlTpTeewC+mTAGM+dKPBI0aygKqSZc8qgwd8xMvOoByABYQANv6IXgA0mHpQAIY0sAiI7AD0AFTGpsBQ-iTAELiR-gCKRNzAvIx2Ws5YOG6kFN7MEIw5eVAA1o2IAEYkoRxcPHzUmjI65GHi9rKeGpOjYRHRsUjsSQlAA